File: server/src/main/java/org/queue/javaapi/message/ByteBufferMessageSet.scala
/*
 * Copyright 2010 LinkedIn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/
package org.queue.javaapi.message

import org.queue.common.ErrorMapping
import org.queue.message.{CompressionCodec, CompressionUtils, Message, MessageAndOffset, MessageSet, NoCompressionCodec}
import org.slf4j.LoggerFactory

import java.nio.ByteBuffer
import java.nio.channels.WritableByteChannel
import scala.jdk.CollectionConverters.CollectionHasAsScala

class ByteBufferMessageSet(private val buffer: ByteBuffer,
                           private val initialOffset: Long = 0L,
                           private val errorCode: Int = ErrorMapping.NoError) extends MessageSet {
  private val logger = LoggerFactory.getLogger(getClass())
  val underlying: org.queue.message.ByteBufferMessageSet = new org.queue.message.ByteBufferMessageSet(buffer,
                                                                                              initialOffset,
                                                                                              errorCode)
  def this(buffer: ByteBuffer) = this(buffer, 0L, ErrorMapping.NoError)

  def this(compressionCodec: CompressionCodec, messages: java.util.List[Message]) {
    this(compressionCodec match {
      case NoCompressionCodec =>
        val buffer = ByteBuffer.allocate(MessageSet.messageSetSize(messages))
        val messageIterator = messages.iterator
        while(messageIterator.hasNext) {
          val message = messageIterator.next
          message.serializeTo(buffer)
        }
        buffer.rewind
        buffer
      case _ =>

        val message = CompressionUtils.compress(messages.asScala, compressionCodec)
        val buffer = ByteBuffer.allocate(message.serializedSize)
        message.serializeTo(buffer)
        buffer.rewind
        buffer
    }, 0L, ErrorMapping.NoError)
  }

  def this(messages: java.util.List[Message]) {
    this(NoCompressionCodec, messages)
  }

  def validBytes: Long = underlying.validBytes

  def serialized():ByteBuffer = underlying.serialized

  def getInitialOffset = initialOffset

  def getBuffer = buffer

  def getErrorCode = errorCode

  override def iterator: Iterator[MessageAndOffset] = new Iterator[MessageAndOffset] {
    val underlyingIterator = underlying.iterator
    override def hasNext: Boolean = {
      underlyingIterator.hasNext
    }
    override def next(): MessageAndOffset = {
      underlyingIterator.next
    }
  }

  override def toString: String = underlying.toString

  def sizeInBytes(): Long = underlying.sizeInBytes

  override def equals(other: Any): Boolean = {
    other match {
      case that: ByteBufferMessageSet =>
        (that canEqual this) && errorCode == that.errorCode && buffer.equals(that.buffer) && initialOffset == that.initialOffset
      case _ => false
    }
  }

  def canEqual(other: Any): Boolean = other.isInstanceOf[ByteBufferMessageSet]

  override def hashCode: Int = 31 * (17 + errorCode) + buffer.hashCode + initialOffset.hashCode

  /** Write the messages in this set to the given channel starting at the given offset byte.
   * Less than the complete amount may be written, but no more than maxSize can be. The number
   * of bytes written is returned */
  override def writeTo(channel: WritableByteChannel, offset: Long, maxSize: Long): Long = {
    offset
  }
}

File: server/src/main/java/org/queue/javaapi/message/ByteBufferMessageSet.java
//package org.queue.javaapi.message;
//
//
//import org.queue.common.ErrorMapping;
//
//
//import org.queue.common.ErrorMapping;
//import org.slf4j.Logger;
//import org.slf4j.LoggerFactory;
//import java.nio.ByteBuffer;
//import java.nio.channels.WritableByteChannel;
//import java.util.List;
//import java.util.ArrayList;
//import java.util.Iterator;
//import java.io.IOException;
//
//public class ByteBufferMessageSet extends MessageSet {
//    private final ByteBuffer buffer;
//    private final long initialOffset;
//    private final int errorCode;
//    private static final Logger logger = LoggerFactory.getLogger(ByteBufferMessageSet.class);
//
//    // 假设org.queue.message.ByteBufferMessageSet存在，并提供必要的方法
//    private final org.queue.message.ByteBufferMessageSet underlying;
//
//    public ByteBufferMessageSet(ByteBuffer buffer, long initialOffset, int errorCode) {
//        this.buffer = buffer;
//        this.initialOffset = initialOffset;
//        this.errorCode = errorCode;
//        this.underlying = new org.queue.message.ByteBufferMessageSet(buffer, initialOffset, errorCode);
//    }
//
//    public ByteBufferMessageSet(ByteBuffer buffer) {
//        this(buffer, 0L, ErrorMapping.NoError);
//    }
//
//    public ByteBufferMessageSet(CompressionType compressionCodec, List<Message> messages) {
//        this(createByteBuffer(compressionCodec, messages), 0L, ErrorMapping.NoError);
//    }
//
//    public ByteBufferMessageSet(List<Message> messages) {
//        this(CompressionType.NONE, messages);
//    }
//
//    public long validBytes() {
//        return underlying.validBytes();
//    }
//
//    public ByteBuffer serialized() {
//        return underlying.serialized();
//    }
//
//    public long getInitialOffset() {
//        return initialOffset;
//    }
//
//    public ByteBuffer getBuffer() {
//        return buffer;
//    }
//
//    public int getErrorCode() {
//        return errorCode;
//    }
//
//    @Override
//    public Iterator<MessageAndOffset> iterator() {
//        return new Iterator<MessageAndOffset>() {
//            private final Iterator<MessageAndOffset> underlyingIterator = underlying.iterator();
//
//            @Override
//            public boolean hasNext() {
//                return underlyingIterator.hasNext();
//            }
//
//            @Override
//            public MessageAndOffset next() {
//                if (!hasNext()) {
//                    throw new NoSuchElementException();
//                }
//                return underlyingIterator.next();
//            }
//        };
//    }
//
//    @Override
//    public String toString() {
//        return underlying.toString();
//    }
//
//    public long sizeInBytes() {
//        return underlying.sizeInBytes();
//    }
//
//    @Override
//    public boolean equals(Object other) {
//        if(other instanceof ByteBufferMessageSet){
//            ByteBufferMessageSet that = (ByteBufferMessageSet) other;
//            return errorCode == that.errorCode &&
//                    initialOffset == that.initialOffset &&
//                    buffer.equals(that.buffer);
//        }else {
//            return false;
//        }
//    }
//
//    @Override
//    public int hashCode() {
//        int result = errorCode;
//        result = 31 * result + buffer.hashCode();
//        result = 31 * result + (int) (initialOffset ^ (initialOffset >>> 32));
//        return result;
//    }
//
//    @Override
//    public long writeTo(WritableByteChannel channel, long offset, long maxSize) throws IOException {
//        // 根据提供的offset和maxSize将消息写入WritableByteChannel
//        // 这里需要实现具体的写入逻辑，下面代码是示例性的
//        long bytesWritten = 0;
//        while (bytesWritten < maxSize && buffer.hasRemaining()) {
//            channel.write(buffer.slice());
//            bytesWritten += buffer.position();
//            buffer.position(0); // Reset position after write
//        }
//        return bytesWritten;
//    }
//
//    private ByteBuffer createByteBuffer(CompressionType compressionCodec, List<Message> messages) {
//        // 根据压缩类型和消息列表创建ByteBuffer
//        // 这里需要实现具体的序列化和压缩逻辑，下面代码是示例性的
//        ByteBuffer buffer = ByteBuffer.allocate(1024); // 假设分配了足够的空间
//        for (Message message : messages) {
//            message.serializeTo(buffer);
//        }
//        buffer.flip();
//        return buffer;
//    }
//}